src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c

```diff
@@ -119,21 +119,23 @@ void PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
 
     while( waitOperation > 0 )
     {
-        PaPulseAudio_Lock( hostapi->mainloop );
-        pa_threaded_mainloop_wait( hostapi->mainloop );
-        PaPulseAudio_UnLock( hostapi->mainloop );
 
+        PaPulseAudio_Lock( hostapi->mainloop );
         localOperationState = pa_operation_get_state( localOperation );
 
-        // No wait if operation have been DONE or CANCELLED
-        if( localOperationState != PA_OPERATION_RUNNING)
+        if( localOperationState == PA_OPERATION_RUNNING )
+        {
+            pa_threaded_mainloop_wait( hostapi->mainloop );
+        }
+        else
         {
+            // Result is DONE or CANCEL
+            PaPulseAudio_UnLock( hostapi->mainloop );
             break;
         }
+        PaPulseAudio_UnLock( hostapi->mainloop );
```

Comment:
Suggestion: the control flow would be easier to follow if this UnLock operation was moved inside the `if(){}` branch

Reply:
I'm not sure whether you can do anything about it, but note that the following still has a race condition:

```
        localOperationState = pa_operation_get_state( localOperation );
        if( localOperationState == PA_OPERATION_RUNNING )
        {
            pa_threaded_mainloop_wait( hostapi->mainloop );
```

Reply:
Sorry I don't see the race condition here a we are in locked state? Could you explain little bit that I can fix it or if it can't then it's just racing.

Reply:
Sorry, I just checked the documentation for `pa_threaded_mainloop_wait` here https://www.freedesktop.org/software/pulseaudio/doxygen/thread-mainloop_8h.html#ac96d19260567d0406350f1b3a14c4bed

It is clear that the lock is released-reacquired by main-loop wait. So presumably there is no race between checking the stream state and waiting for it to change.